lib/saluki-components/src/transforms/name_filter/mod.rs

```diff
@@ -0,0 +1,238 @@
+use serde::Deserialize;
+use tokio::select;
+use tracing::{debug, error};
+
+#[derive(Deserialize)]
+/// Name Transform.
+pub struct NameFilterConfiguration {
+    #[serde(default, rename = "statsd_metric_namespace")]
+    metric_prefix: String,
+
+    #[serde(
+        default = "default_metric_prefix_blacklist",
+        rename = "statsd_metric_namespace_blacklist"
+    )]
+    metric_prefix_blacklist: Vec<String>,
+
+    #[serde(default, rename = "statsd_metric_blocklist")]
+    metric_blocklist: Vec<String>,
+
+    #[serde(default, rename = "statsd_metric_blocklist_match_prefix")]
+    metric_blocklist_match_prefix: bool,
+}
+
+fn default_metric_prefix_blacklist() -> Vec<String> {
+    vec![
+        "datadog.agent".to_string(),
+        "datadog.dogstatsd".to_string(),
+        "datadog.process".to_string(),
+        "datadog.trace_agent".to_string(),
+        "datadog.tracer".to_string(),
+        "activemq".to_string(),
+        "activemq_58".to_string(),
+        "airflow".to_string(),
+        "cassandra".to_string(),
+        "confluent".to_string(),
+        "hazelcast".to_string(),
+        "hive".to_string(),
+        "ignite".to_string(),
+        "jboss".to_string(),
+        "jvm".to_string(),
+        "kafka".to_string(),
+        "presto".to_string(),
+        "sidekiq".to_string(),
+        "solr".to_string(),
+        "tomcat".to_string(),
+        "runtime".to_string(),
+    ]
+}
+
+impl NameFilterConfiguration {
+    /// Creates a new `NameFilterConfiguration` from the given configuration.
+    pub fn from_configuration(config: &GenericConfiguration) -> Result<Self, GenericError> {
+        Ok(config.as_typed()?)
+    }
+}
+
+#[async_trait]
+impl TransformBuilder for NameFilterConfiguration {
+    fn input_data_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn outputs(&self) -> &[OutputDefinition] {
+        static OUTPUTS: &[OutputDefinition] = &[OutputDefinition::default_output(DataType::Metric)];
+        OUTPUTS
+    }
+
+    async fn build(&self, _: ComponentContext) -> Result<Box<dyn Transform + Send>, GenericError> {
+        Ok(Box::new(NameFilter {
+            metric_prefix: self.metric_prefix.clone(),
+            metric_prefix_blacklist: self.metric_prefix_blacklist.clone(),
+            blocklist: Blocklist::new(&self.metric_blocklist, self.metric_blocklist_match_prefix),
+        }))
+    }
+}
+
+impl MemoryBounds for NameFilterConfiguration {
+    fn specify_bounds(&self, builder: &mut MemoryBoundsBuilder) {
+        // Capture the size of the heap allocation when the component is built.
+        builder.minimum().with_single_value::<NameFilter>();
+    }
+}
+
+#[derive(Debug)]
+struct Blocklist {
+    data: Vec<String>,
+    match_prefix: bool,
+}
+
+impl Blocklist {
+    fn new(data: &[String], match_prefix: bool) -> Self {
+        let mut data = data.to_owned();
+        data.sort();
+
+        if match_prefix && !data.is_empty() {
+            let mut i = 0;
```

Comment:
I think a more important code comment to have here is that by doing this, we're only keeping unique prefixes: that's the crux of the filtering we're doing here and how it's then used in `test`.